presto-main/src/test/java/com/facebook/presto/operator/TestPartitionedOutputOperator.java

```diff
@@ -0,0 +1,311 @@
+                    .createOperator(createDriverContext());
+        }
+        else {
+            operatorFactory = new PartitionedOutputOperator.PartitionedOutputFactory(
+                    partitionFunction,
+                    ImmutableList.of(0),
+                    ImmutableList.of(Optional.empty(), Optional.empty()),
+                    false,
+                    OptionalInt.empty(),
+                    buffer,
+                    PARTITION_MAX_MEMORY);
+            return (PartitionedOutputOperator) operatorFactory
+                    .createOutputOperator(0, new PlanNodeId("plan-node-0"), TYPES, Function.identity(), serdeFactory)
+                    .createOperator(createDriverContext());
+        }
+    }
+
+    private PartitionedOutputBuffer createPartitionedOutputBuffer()
+    {
+        OutputBuffers buffers = createInitialEmptyOutputBuffers(PARTITIONED);
+        for (int partition = 0; partition < PARTITION_COUNT; partition++) {
+            buffers = buffers.withBuffer(new OutputBuffers.OutputBufferId(partition), partition);
+        }
+        PartitionedOutputBuffer buffer = createPartitionedBuffer(
+                buffers.withNoMoreBufferIds(),
+                new DataSize(Long.MAX_VALUE, BYTE)); // don't let output buffer block
+        buffer.registerLifespanCompletionCallback(ignore -> {});
+
+        return buffer;
+    }
+
+    private Page createSimplePage()
+    {
+        return new Page(buildBigIntBlock(POSITION_COUNT, false));
+    }
+
+    private Page createPageWithDictionary()
+    {
+        return new Page(buildDictionaryBlock(buildBigIntBlock(POSITION_COUNT, false), POSITION_COUNT));
+    }
+
+    private Page createPageWithRunLength()
+    {
+        return new Page(buildRleBlock(buildBigIntBlock(POSITION_COUNT, false), POSITION_COUNT));
+    }
+
+    private Page createPageWithNullBlock()
+    {
+        Block[] blocks = new Block[2];
+        blocks[0] = new RunLengthEncodedBlock(BIGINT.createBlockBuilder(null, 1).appendNull().build(), POSITION_COUNT);
+        blocks[1] = buildBigIntBlock(POSITION_COUNT, false);
+
+        return new Page(POSITION_COUNT, blocks);
+    }
+
+    private Page createPageWithDictionaryAndNullBlock()
+    {
+        Block[] blocks = new Block[2];
+        blocks[0] = new RunLengthEncodedBlock(BIGINT.createBlockBuilder(null, 1).appendNull().build(), POSITION_COUNT);
+        blocks[1] = buildDictionaryBlock(buildBigIntBlock(POSITION_COUNT, false), POSITION_COUNT);
+
+        return new Page(POSITION_COUNT, blocks);
+    }
+
+    private Page createPageWithRunLengthAndNullBlock()
+    {
+        Block[] blocks = new Block[2];
+        blocks[0] = new RunLengthEncodedBlock(BIGINT.createBlockBuilder(null, 1).appendNull().build(), POSITION_COUNT);
+        blocks[1] = buildRleBlock(buildBigIntBlock(POSITION_COUNT, false), POSITION_COUNT);
+
+        return new Page(POSITION_COUNT, blocks);
+    }
+
+    private DriverContext createDriverContext()
```

Comment:
inline this method; it is used only once

Reply:
This is actually used twice.

Reply:
I should have been clearer. 
- `createDriverContext` is used only in `createPartitionedOutputOperator`. 
- It is used in two branches of a single if-else block. 
- There is `createPartitionedOutputBuffer` that could have been used the same way, but it is not. Instead, a local variable outside of if-else block hold the value of `createPartitionedOutputBuffer()` and if used in both branches.
- To be consistent, another local variable should be used to hold the value of `createDriverContext()` and used in both branches. 
- Once you make that change, there will be only single usage of `createDriverContext` and single usage of `createPartitionedOutputBuffer`. 
- `createPartitionedOutputOperator` is relatively small and therefore the preference is to inline both `createDriverContext` and `createPartitionedOutputBuffer`. This makes it easier to read the code as the reader doesn't need to jump back and forth between different small functions.

Here is what we could have:

```
    private PartitionedOutputOperator createPartitionedOutputOperator(boolean shouldReplicate)
    {
        PartitionFunction partitionFunction = new LocalPartitionGenerator(new InterpretedHashGenerator(ImmutableList.of(BIGINT), new int[] {0}), PARTITION_COUNT);
        PagesSerdeFactory serdeFactory = new PagesSerdeFactory(new BlockEncodingManager(new TypeRegistry()), false);

        DriverContext driverContext = TestingTaskContext.builder(EXECUTOR, SCHEDULER, TEST_SESSION)
                .setMemoryPoolSize(MAX_MEMORY)
                .setQueryMaxTotalMemory(MAX_MEMORY)
                .build()
                .addPipelineContext(0, true, true, false)
                .addDriverContext();

        OutputBuffers buffers = createInitialEmptyOutputBuffers(PARTITIONED);
        for (int partition = 0; partition < PARTITION_COUNT; partition++) {
            buffers = buffers.withBuffer(new OutputBuffers.OutputBufferId(partition), partition);
        }
        PartitionedOutputBuffer buffer = new PartitionedOutputBuffer(
                "task-instance-id",
                new StateMachine<>("bufferState", SCHEDULER, OPEN, TERMINAL_BUFFER_STATES),
                buffers.withNoMoreBufferIds(),
                new DataSize(Long.MAX_VALUE, BYTE),
                () -> new SimpleLocalMemoryContext(newSimpleAggregatedMemoryContext(), "test"),
                SCHEDULER);
        buffer.registerLifespanCompletionCallback(ignore -> {});

        PartitionedOutputOperator.PartitionedOutputFactory operatorFactory;
        if (shouldReplicate) {
            operatorFactory = new PartitionedOutputOperator.PartitionedOutputFactory(
                    partitionFunction,
                    ImmutableList.of(0),
                    ImmutableList.of(Optional.empty()),
                    true,
                    OptionalInt.of(0),
                    buffer,
                    PARTITION_MAX_MEMORY);
            return (PartitionedOutputOperator) operatorFactory
                    .createOutputOperator(0, new PlanNodeId("plan-node-0"), REPLICATION_TYPES, Function.identity(), serdeFactory)
                    .createOperator(driverContext);
        }
        else {
            operatorFactory = new PartitionedOutputOperator.PartitionedOutputFactory(
                    partitionFunction,
                    ImmutableList.of(0),
                    ImmutableList.of(Optional.empty(), Optional.empty()),
                    false,
                    OptionalInt.empty(),
                    buffer,
                    PARTITION_MAX_MEMORY);
            return (PartitionedOutputOperator) operatorFactory
                    .createOutputOperator(0, new PlanNodeId("plan-node-0"), TYPES, Function.identity(), serdeFactory)
                    .createOperator(driverContext);
        }
    }
```